fix: prevent portal from distorting on viewport resize.

wrangelvid opened the pull request:
When changing the browser window with a Portal component, it is distorting. This is due to a missmatch in the aspect ratio of the FBO and the camera used to render the FBA, which inherits the aspect ratio from the main canvas upon window resize.

To fix this, we set the aspect ratio to 1 before rendering to the FBO and reset it afterwards.

To test this, try the following setup and resize your browser window:

```javascript
<Canvas>
    <ambientLight intensity={0.5} />
     <Fullscreen>
         <Portal width={200} aspectRatio={1}>
              <mesh>
                   <boxGeometry />
                    <meshBasicMaterial color="red" />
               </mesh>
        </Portal>
    </Fullscreen>
</Canvas>
```

Note that there might be a better implementation that resizes the FBO dynamically rather than the updating the camera in useFrame.

wrangelvid commented:
Note, this only a problem if no camera is provided to the portal.

bbohlender commented:
thanks for the PR :)

Concept looks good, but it only works for PerspectiveCameras and you're hard coding the aspect ratio. I'll check how R3F updates the camera based on the size in the state on render.

wrangelvid commented:
> Concept looks good, but it only works for PerspectiveCameras and you're hard coding the aspect ratio. I'll check how R3F updates the camera based on the size in the state on render.

In [store.ts ](https://github.com/pmndrs/react-three-fiber/blob/6b1029f8f54cc0951f8b8b912da8e792b663a73d/packages/fiber/src/core/store.ts#L343) they call [updateCamera](https://github.com/pmndrs/react-three-fiber/blob/6b1029f8f54cc0951f8b8b912da8e792b663a73d/packages/fiber/src/core/utils.ts#L460) whenever the screen resizes, which is why we are seeing the portal distorting.

I also did some [digging](https://github.com/pmndrs/react-three-fiber/blob/6b1029f8f54cc0951f8b8b912da8e792b663a73d/example/src/demos/Portals.tsx#L52). They avoid the headache by creating a new camera in the `Portals` demo by default.

bbohlender commented:
Seems reasonable, so let's create a new camera by default and use the same code (if possible directly the R3F updateCamera function) to update the camera before rendering.
Do you want to make these modifications to your PR or should I do them?

wrangelvid commented:
Great! Let me take a crack at this later today 🙂

wrangelvid commented:
Unfortunately, the `updateCamera` function is not exposed by r3f, but since we know we create a PerspectiveCamera, it should be fine.

bbohlender commented:
Yes, but the camera in the state can be overwritten by the contents of the Portal. Therefore, you cannot assume that the camera injected into the state as the default, is the camera the will be used to render the scene.

bbohlender commented on packages/react/src/portal.tsx:
```diff
@@ -68,7 +74,7 @@ export const Portal: (props: PortalProperties & RefAttributes<ComponentInternals
             </ChildrenToFBO>,
             vScene,
-            injectState,
+            { ...injectState, camera: camera },
```
You're recreating the state on every react re-render

wrangelvid replied:
Are you suggesting we should replace the line with something like this:
`useMemo(() => ({ ...injectState, camera: camera }), [injectState, camera]),`
Or do you envision something better ? :)

bbohlender replied:
you can just create the camera inside the useMemo for creating the "injectState" which is semantically more cleaner (creating the injectState in one place) and more performant (1 instead of 3 useMemo)

wrangelvid replied:
Clean! I like it.

bbohlender commented on packages/react/src/portal.tsx:
```diff
@@ -48,6 +50,10 @@ export const Portal: (props: PortalProperties & RefAttributes<ComponentInternals
           const [width, height] = size.value
           const dpr = store.getState().viewport.dpr
           renderTarget.setSize(width * dpr, height * dpr)
+          injectState.camera!.aspect = width / height;
```
As commented before, this should be done using the store.getState().camera (not literally and not here because you cannot get the store from inside the portal here)
Therefore, imo the best approach is to get the store `useStore` inside `ChildrenToFBO` and subscribe to it inside a useEffect and apply the size of the state form the store to the camera of the state from the state

wrangelvid replied:
I see your point, but initally we need to set the position and rotation of the portal camera to something non-zero. The r3f Portals example sets it to the scenere cameras position and rotation, which we only have access here since injectState will override it until we get to useFrame. So I think we need both. Here, we should set the position and rotation of the camera. Then in useFrame, we can update the aspect ratio based on the size.

bbohlender replied:
I was not referring the the position code. That part is fine although we could think about directly setting the position when creating the camera inside the memo. I am also not sure copying the camera's transformation is better then just setting it to something non zero. Copying could lead to unexpected behavior, e.g. "in this example it works" but here it soundenly doesn't work"

bbohlender commented:
@wrangelvid sorry for the delay
I've just check it out locally and everything seems to work fine (I've added a manual test to examples/uikit).
I have 2 small comments to the code, other than that, it seems correct for me. Can you elaborate on what is not working in your tests? :)

bbohlender commented on packages/react/src/portal.tsx:
```diff
@@ -27,6 +27,12 @@ export const Portal: (props: PortalProperties & RefAttributes<ComponentInternals
           events: { compute: uvCompute.bind(null, imageRef), priority: eventPriority },
           size: { width: 1, height: 1, left: 0, top: 0 },
+          // We have our own camera in here, separate from the main scene.
+          camera: (() => {
```
nit pick: I thinks its easier to read if you put the camera initialization at the start of the function
e.g. `() => { const camera = ...; return { ..., camera } }`

bbohlender commented on packages/react/src/portal.tsx:
```diff
@@ -123,6 +129,19 @@ function ChildrenToFBO({
   })
+
+  useEffect(() => {
+    const unsubscribe = store.subscribe((state) => {
```
3 small things:
1. you can directly return the result of the store.subscribe
2. its better to only update the projection matrix if aspect or camera changed (e.g. `(state, prevState) => state.camera != prevState.camera ....`)
3. your code only works with a PerspectiveCamera, use the same code as R3F to make it also work with a OrthographicCamera

wrangelvid commented:
> @wrangelvid sorry for the delay I've just check it out locally and everything seems to work fine (I've added a manual test to examples/uikit). I have 2 small comments to the code, other than that, it seems correct for me. Can you elaborate on what is not working in your tests? :)

Back to the keys :) I saw your modified test cases, happy to add some tests in there as well. I have been testing externally since I had a few import issues with setting up the dev environment locally. Are there developer guidelines/instructions for UIKit you can point me to?

bbohlender commented:
Yeah sure, I hope the ones in here are up to date. Let me know if you have any issue
https://github.com/pmndrs/uikit/blob/main/CONTRIBUTING.md

wrangelvid commented:
> @wrangelvid sorry for the delay I've just check it out locally and everything seems to work fine (I've added a manual test to examples/uikit). I have 2 small comments to the code, other than that, it seems correct for me. Can you elaborate on what is not working in your tests? :)

I added two more portals to illustrate the issue. To me all the cubes look exactly the same. The provided camera doesn't affect it the portal scene visually.

> Yeah sure, I hope the ones in here are up to date. Let me know if you have any issue

Wow, I really missed that ALL_CAPS file :-)

bbohlender commented:
The problem is actually inside `createPortal` from R3F because it overrides the state changes made inside the scene with the initial state each time the parent state changes.
I'll try to implement a custom create portal here for now in a sec :)

bbohlender commented:
@wrangelvid building a  portal implementation that "correctly" merges the outer and the inner state took me longer then hoped but it should be complete now

wrangelvid commented:
> @wrangelvid building a portal implementation that "correctly" merges the outer and the inner state took me longer then hoped but it should be complete now

Just verified the test cases. Works like a charm!

wrangelvid commented:
@bbohlender Is there a regular deployment cycle to the central package repositories? In other words, when can we expect this to be in 0.3.5 :)

bbohlender commented:
I am doing releases as new features are comming in. I wanted to get PR #68 into the next release so it will be tomorrow if that works for you guys :)
